### PR TITLE
fix mem leak on error path and improve error handling

### PR DESCRIPTION
**Pull request by yuchi1989**
fix mem leak on error path and improve error handling in some test files

### CODE REVIEW

File: test/ssltest_old.c [hunk 1]
```diff
@@ -391,6 +391,9 @@ static int cb_server_alpn(SSL *s, const unsigned char **out,
      */
     alpn_selected = OPENSSL_malloc(*outlen);
     memcpy(alpn_selected, *out, *outlen);
+    if(*out != NULL) {
+        OPENSSL_free(*out);
```
Comment (bernd-edlinger) [inline test/ssltest_old.c]: no need for checking *out != NULL here

File: test/bntest.c [hunk 2]
```diff
@@ -185,6 +185,7 @@ static int test_sub()
 {
     BIGNUM *a, *b, *c;
     int i;
+    int st = 0;
 
     a = BN_new();
```
Comment (bernd-edlinger) [inline test/bntest.c]: how about checking that BN_new returns something here?

File: test/ssltest_old.c (hunk 1 above)
Comment (richsalz) [inline test/ssltest_old.c]: yes, in fact *remove* the check.

**Comment by richsalz** [comment]
@bernd-edlinger has good comments.  what you should do is make the changes, do "commit --amend" to update the single  commit, and then go "git push -f github HEAD" to push the updated commit.  Don't need to build up a set of commits which will then get squashed.  Make sense?

File: test/ssltest_old.c (hunk 1 above)
Comment (yuchi1989) [inline test/ssltest_old.c]: Okay. I see. I will remove the check.

**Comment by yuchi1989** [comment]
Thank you. I did not know that a commit could be updated.

File: test/bntest.c (hunk 2 above)
Comment (yuchi1989) [inline test/bntest.c]: Do you mean for all the similar statements in this file like `a = BN_new();`. we should check them by `if (a == NULL) goto err;`

Comment (richsalz) [inline test/bntest.c]: or do all the BN_new calls and then a single if to test if a b or c is NULL.
Or, for tests, do not worry very much.  The test will crash and that's useful :)

Comment (yuchi1989) [inline test/bntest.c]: I see. I will do as you suggest.

**Comment by yuchi1989** [comment]
Is there something wrong with the patches? What does the failure indicate?

**Comment by levitte** [comment]
> Is there something wrong with the pr? What does the failure indicate?

No fault of yours.  The error has been fixed in master, so if you just rebase on a fresher master checkout, things will be fine again.

**Comment by richsalz** [comment]
So if you do this:
```
git checkout $YOURBRANCHNAME
git rebase master
git push -f github
```
That will rebase and push it to github.

**Comment by yuchi1989** [comment]
Okay. Thank you.

File: test/bntest.c [hunk 3]
```diff
@@ -372,16 +388,18 @@ static int test_modexp_mont5()
     n = BN_new();
     c = BN_new();
     mont = BN_MONT_CTX_new();
+    if (a == NULL || p == NULL || m == NULL || d == NULL || e == NULL || b == NULL || n == NULL || c == NULL || mont == NULL)
```
Comment (mattcaswell) [inline test/bntest.c]: This line is too long. Max 80 chars per line.

File: test/evp_test.c [hunk 4]
```diff
@@ -1389,8 +1389,11 @@ static int pkey_test_init(struct evp_test *t, const char *name,
     kdata->keyop = keyop;
     t->data = kdata;
     kdata->ctx = EVP_PKEY_CTX_new(pkey, NULL);
-    if (!kdata->ctx)
+    if (!kdata->ctx) {
```
Comment (mattcaswell) [inline test/evp_test.c]: We should change this to `if (kdata->ctx == NULL)`

File: test/bntest.c (hunk 3 above)
Comment (yuchi1989) [inline test/bntest.c]: Sure. I have changed them.

File: test/evp_test.c (hunk 4 above)
Comment (yuchi1989) [inline test/evp_test.c]: Thank you. I have changed it as you suggest.

**Comment by richsalz** [comment]
I think this became outdated when we converted to the new test framework, so I am going to close this.  Thanks for your contribution!  If I'm wrong, and there is still work to do in test/bntest.c, please re-open this or submit a new PR.

**Pull request closed without merging**